Add result_type field for http_response input

froth opened the pull request:
Fix for https://github.com/influxdata/telegraf/issues/2784 & https://github.com/influxdata/telegraf/issues/2624.

Does return measurement data even if no http connection can be established.

This is based on pull request https://github.com/influxdata/telegraf/pull/2813, as it depends on the fixes.

### Required for all PRs:

- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

phemmer commented on plugins/inputs/http_response/http_response.go:
```diff
@@ -130,15 +130,21 @@ func (h *HTTPResponse) httpGather() (map[string]interface{}, error) {
 		if h.FollowRedirects {
-			return nil, err
+			log.Printf("E! Failed to open HTTP connection: %s", err)
+			return fields, nil
```
Why are we directly logging the error instead of returning it?

froth replied:
The other option would be to give the the fields and the errors one level up and logging them. I thought about that and my reasoning is the following:
Currently we either return the fields or the error and this would break this pattern. In my opinion this is a little easier to read. However I don't have a strong opinion on it. I can change it, if it's important for you.

phemmer replied:
The reason I bring it up is 2 things. Doing it this way results in the log messages not having context. Meaning that if you have multiple inputs defined, you don't know which one is generating the error. The other problem is that I think we have an open issue somewhere to track statistics on per-input error counts, which requires errors going through `Accumulator.AddError()`.

danielnelson commented:
Check out my comment on #2784.

froth commented:
@phemmer I amended my commit to reflect your comments. Can you have a look?

froth commented:
I updated the pull request to reflect the discussion in https://github.com/influxdata/telegraf/issues/2784

danielnelson commented on plugins/inputs/http_response/http_response.go:
```diff
@@ -130,15 +131,24 @@ func (h *HTTPResponse) httpGather() (map[string]interface{}, error) {
+		fields["connection_failed"] = 1
 		if h.FollowRedirects {
-			return nil, err
+			return fields, err
```
In general, returning an error indicates that the value is not valid, if we are returning fields then we shouldn't have an error.  Later on, we will need to make sure we don't call AddFields if it is nil.

danielnelson commented on plugins/inputs/http_response/http_response.go:
```diff
@@ -130,15 +131,24 @@ func (h *HTTPResponse) httpGather() (map[string]interface{}, error) {
+
 	if err != nil {
+		fields["http_response_code"] = -1
+		fields["response_string_match"] = 0
```
There was no response so these fields should not be set.

danielnelson commented on plugins/inputs/http_response/http_response.go:
```diff
@@ -130,15 +131,24 @@ func (h *HTTPResponse) httpGather() (map[string]interface{}, error) {
+		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+			fields["response_timeout"] = time.Since(start).Seconds()
+		} else {
+			fields["response_time"] = time.Since(start).Seconds()
```
I think this goes away too, there was no response.

froth commented:
Thanks for your comments. I updated the pull request to reflect them :)

phemmer commented on plugins/inputs/http_response/http_response.go:
```diff
@@ -148,6 +154,7 @@ func (h *HTTPResponse) httpGather() (map[string]interface{}, error) {
 
 	fields["response_time"] = time.Since(start).Seconds()
 	fields["http_response_code"] = resp.StatusCode
+	fields["connection_failed"] = 0
```
My thoughts:
It looks like we have 3 different fields that behave similarly, connection_failed, response_timeout, and response_string_match. This feels wrong to me. All three fields are exclusive of each other. You'll never have more than one of them true. So it seems like we should instead have one field which indicates the result. Something like `result_type` with the values `connection failed`/`timeout`/`response match`.

If we do keep them separate fields, I would at least argue that they should be booleans, not ints.

danielnelson replied:
One consequence of combining them is that we would not be able to store how long the timeout took, but maybe we don't care about this since it is basically set by hand.

Interesting idea to work match/mismatch in, do you think we would leave the field off if no match string is provided?  Perhaps we have a success result_type?

I was letting the int slide here because this plugin already has the response match field with one, but yeah in general we should use booleans.

froth replied:
I think it is a good idea to always return a status whether the connection failed or not. If someone (i.e. me) wants to gather the latest value of the status and it is only returned when an error occured the value is always "error". 

Do we have a plan if we want to combine the fields and if so: how?

I can change the int fields to boolean if you want, however that would mean a breaking change for the response_match field.

froth commented:
@danielnelson, @phemmer 
I fixed the return error part. 

Do we have a plan for the "connection_failed" field?
The way I have implemented it would fix the concrete problem I have. If you have input what can be done better I am willing to improve it further. 

Currently I have the feeling that we are kind of stuck in the discussion of how to deal with the flag-fields. I think it is an important discussion and I appreciate that you take it seriously. But to be honest my current main concern is "getting the bug fixed soon, so that my alerting system is reliable".

danielnelson commented:
As usual, @phemmer's solution is quite a bit better than the one i suggested earlier, so how about:

- Forget about the `response_timeout` field for now, it's not clear if we need it and we can always add it later.
- Add `result_type` field with possible values `connection_failed`, `timeout`, `response_string_mismatch`, `success`.  This will replace the `connection_failed` field we had earlier and should be more extensible.

Even with this change, it would probably be a good idea to make sure you are alerting if the measurement you are looking for stops being reported.

froth commented:
Done :)

Please have a look

danielnelson commented:
Leave the old response_string_mismatch field around so we don't break anything.

froth commented:
@danielnelson done :)

danielnelson commented:
@froth It's in!  Thanks and sorry about all the backtracking while we figured out what to do.
